contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java

```diff
@@ -40,27 +51,107 @@
  * @since 0.3.2
  */
-public abstract class HBaseScanOperator<T> extends HBaseInputOperator<T>
+public abstract class HBaseScanOperator<T> extends HBaseInputOperator<T> implements Operator.ActivationListener<Context>
 {
+  public static final int DEF_HINT_SCAN_LOOKAHEAD = 2;
+  public static final int DEF_QUEUE_SIZE = 1000;
+  public static final int DEF_SLEEP_MILLIS = 10;
+
+  private String startRow;
+  private String endRow;
+  private String lastReadRow;
+  private int hintScanLookahead = DEF_HINT_SCAN_LOOKAHEAD;
+  private int queueSize = DEF_QUEUE_SIZE;
+  private int sleepMillis = DEF_SLEEP_MILLIS;
+  private Queue<Result> resultQueue;
+  private boolean threadFailed = false;
+
+  @AutoMetric
+  protected long tuplesRead;
+
+  // Transients
+  protected transient Scan scan;
+  protected transient ResultScanner scanner;
+  protected transient Thread readThread;
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    super.setup(context);
+    resultQueue = Queues.newLinkedBlockingQueue(queueSize);
+  }
+
+  @Override
+  public void activate(Context context)
+  {
+    startReadThread();
+  }
+
+  protected void startReadThread()
+  {
+    try {
+      scan = operationScan();
+      scanner = getStore().getTable().getScanner(scan);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    readThread = new Thread(new Runnable() {
+      @Override
+      public void run()
+      {
+        try {
+          Result result;
+          while ((result = scanner.next()) != null) {
+            while (!resultQueue.offer(result)) {
+              Thread.sleep(sleepMillis);
+            }
+          }
+        } catch (Exception e) {
+          logger.debug("Exception in fetching results {}", e.getMessage());
+          threadFailed = true;
+          throw new RuntimeException(e);
+        } finally {
+          scanner.close();
+        }
+      }
+    });
+    readThread.start();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    tuplesRead = 0;
+  }
 
   @Override
   public void emitTuples()
   {
+    if (!readThread.isAlive() && threadFailed) {
+      throw new RuntimeException("Exception in scan thread");
+    }
     try {
-      HTable table = getTable();
-      Scan scan = operationScan();
-      ResultScanner scanner = table.getScanner(scan);
-      for (Result result : scanner) {
-        //KeyValue[] kvs = result.raw();
-        //T t = getTuple(kvs);
-        T t = getTuple(result);
-        outputPort.emit(t);
+      Result result = resultQueue.poll();
+      if (result == null) {
+        return;
+      }
+      T tuple = getTuple(result);
+      if (tuple != null) {
+        outputPort.emit(tuple);
+        tuplesRead++;
       }
     } catch (Exception e) {
-      e.printStackTrace();
+      throw new RuntimeException(e);
     }
   }
 
+  @Override
+  public void deactivate()
+  {
+    readThread.interrupt();
```

Comment:
Please don't interrupt I would suggest. Have a variable in thread which is set to true for running and false for existing. Then wait for the operation to finish.